frontend/src/api/prometheus/kservePerformanceMetrics.ts

```diff
@@ -0,0 +1,199 @@
+
+  const inferenceLatency = useQueryRangeResourceData(
+    active,
+    metricsDef.queries[0].query,
+    endInMs,
+    timeframe,
+    defaultResponsePredicate,
+    namespace,
+  );
+
+  const requestLatency = useQueryRangeResourceData(
+    active,
+    metricsDef.queries[1].query,
+    endInMs,
+    timeframe,
+    defaultResponsePredicate,
+    namespace,
+  );
+
+  const data = React.useMemo(
+    () => ({
+      inferenceLatency,
+      requestLatency,
+    }),
+    [inferenceLatency, requestLatency],
+  );
+
+  return useAllSettledContextResourceData(data);
+};
+
+type CpuUsageData = {
+  data: {
+    cpuUsage: PendingContextResourceData<PrometheusQueryRangeResultValue>;
+  };
+  refreshAll: () => void;
+};
+
+export const useFetchKserveCpuUsageData = (
+  metricsDef: KserveMetricGraphDefinition,
+  timeframe: TimeframeTitle,
+  endInMs: number,
+  namespace: string,
+): CpuUsageData => {
+  const active = useIsAreaAvailable(SupportedArea.K_SERVE_METRICS).status;
+
+  const cpuUsage = useQueryRangeResourceData(
+    active,
+    metricsDef.queries[0].query,
+    endInMs,
+    timeframe,
+    defaultResponsePredicate,
+    namespace,
+  );
+
+  const data = React.useMemo(
+    () => ({
+      cpuUsage,
+    }),
+    [cpuUsage],
+  );
+
+  return useAllSettledContextResourceData(data);
+};
+
+type MemoryUsageData = {
+  data: {
+    memoryUsage: PendingContextResourceData<PrometheusQueryRangeResultValue>;
+  };
+  refreshAll: () => void;
+};
+
+export const useFetchKserveMemoryUsageData = (
+  metricsDef: KserveMetricGraphDefinition,
+  timeframe: TimeframeTitle,
+  endInMs: number,
+  namespace: string,
+): MemoryUsageData => {
+  const active = useIsAreaAvailable(SupportedArea.K_SERVE_METRICS).status;
+
+  const memoryUsage = useQueryRangeResourceData(
+    active,
+    metricsDef.queries[0].query,
+    endInMs,
+    timeframe,
+    defaultResponsePredicate,
+    namespace,
+  );
+
+  const data = React.useMemo(
+    () => ({
+      memoryUsage,
+    }),
+    [memoryUsage],
+  );
+
+  return useAllSettledContextResourceData(data);
+};
+
+const useAllSettledContextResourceData = <
+  T,
+  U extends Record<string, PendingContextResourceData<T>>,
+>(
+  data: U,
+) => {
+  const refreshAll = React.useCallback(() => {
+    Object.values(data).forEach((x) => x.refresh());
+  }, [data]);
+
+  const result = React.useMemo(
+    () => ({
+      data,
+      refreshAll,
+    }),
+    [data, refreshAll],
+  );
+
+  // store the result in a reference and only update the reference so long as there are no pending queries
+  //TODO: this creates a bug where the first render may contain a mix of pending / non pending data, causing a flash.
+  //      need to take a default empty value to return first time.
+  const resultRef = React.useRef(result);
+
+  if (!Object.values(result.data).some((x) => x.pending)) {
+    resultRef.current = result;
+  }
+
+  return resultRef.current;
```

Comment:
Do you want all data to be `pending` if any one piece of data is pending? If so, return a new data structure with `pending` set to the overall pending state:

```suggestion
  return React.useMemo(() => {
    const pending = Object.values(data).some((x) => x.pending);
    return {
      data: Object.keys(data).reduce((acc, d) => ({ ...acc, [d]: { ...data[d], pending } }), {
        ...data,
      }),
      refreshAll,
    };
  }, [data, refreshAll]);
```

Reply:
This doesn't fix the issue at hand - chiefly that you can get data that's "pending" returned in the ref - I've updated this to add a default empty state to initially populate the ref with, this fixes the bug.

My solution doesn't win any points for style I imagine, so happy to take any suggestions there :)

Reply:
Yeah, for posterity - I think the better way of implementing this, is to recreate `usePrometheusQueryRange` to enable it to take a batch of queries and use `Promise.all` to make the update atomic. Ran out of time though, I'll try to improve this or log tech debt.

Reply:
I guess my assumption of what you were trying to do was wrong. I'll need more clarity before being able to help further.